spec/services/products_renderer_spec.rb

```diff
@@ -0,0 +1,117 @@
+require 'spec_helper'
+
+describe ProductsRenderer do
+  let(:distributor) { create(:distributor_enterprise) }
+  let(:order_cycle) { create(:simple_order_cycle, distributors: [distributor]) }
+  let(:exchange) { order_cycle.exchanges.to_enterprises(distributor).outgoing.first }
+  let(:customer) { create(:customer) }
+  let(:pr) { ProductsRenderer.new(distributor, order_cycle, customer) }
+
+  describe "sorting" do
+    let(:t1) { create(:taxon) }
+    let(:t2) { create(:taxon) }
+    let!(:p1) { create(:product, name: "abc", primary_taxon_id: t2.id) }
+    let!(:p2) { create(:product, name: "def", primary_taxon_id: t1.id) }
+    let!(:p3) { create(:product, name: "ghi", primary_taxon_id: t2.id) }
+    let!(:p4) { create(:product, name: "jkl", primary_taxon_id: t1.id) }
+
+    before do
+      exchange.variants << p1.variants.first
+      exchange.variants << p2.variants.first
+      exchange.variants << p3.variants.first
+      exchange.variants << p4.variants.first
+    end
+
+    it "sorts products by the distributor's preferred taxon list" do
+      allow(distributor).to receive(:preferred_shopfront_taxon_order) { "#{t1.id},#{t2.id}" }
+      products = pr.send(:products)
```

Comment:
I know this a lot to test in this class but unit tests only cover the public API of a class. If we are not confident about this method is because the class is too big and should be split into smaller ones instead. 

Would you agree on creating an issue for it?

Reply:
These tests were mostly moved across from an existing spec, so I didn't spend a long time rewriting every line to make it prettier. The tests here were pre-existing, and I didn't want to delete them.

Given that this service is incredibly important, I think there is some value in unit testing these private methods, and not just the public interface. What do you think?

Reply:
I'm fine if these were moved but tests that check private methods are not pretty or ugly they simply cost money as private methods are less stable than a public API but also are a symptom of underlying complexity. They are a code smell.

That's why I would ship this but create an issue to work on it right away.